nupic/research/frameworks/pytorch/imagenet/imagenet_tune.py

```diff
@@ -293,11 +289,39 @@ def _stop(self):
     def _process_config(self, config):
         pass
 
-    def _process_result(self, result):
+    @abc.abstractmethod
+    def _run_iteration(self):
+        """Run one iteration of the experiment"""
+        raise NotImplementedError
+
+    def _process_result(self, result, pre_experiment_result=None):
         pass
 
 
-class SigOptImagenetTrainable(ImagenetTrainable):
+class SupervisedTrainable(BaseTrainable):
+    """
+    Trainable class used to train supervised machine learning experiments
+    with ray.
+    """
+
+    def _run_iteration(self):
+        """Run one epoch of training on each process."""
+        status = []
+        for w in self.procs:
+            status.append(w.run_epoch.remote())
+
+        # Wait for remote functions and check for errors
+        if ray_utils.check_for_failure(status):
+            return ray.get(status)
+
+    def _process_result(self, result, pre_experiment_result=None):
+
+        # Aggregate initial validation results (before any training).
+        if pre_experiment_result is not None:
+            result["extra_val_results"].insert(0, (0, pre_experiment_result))
+
+
+class SigOptImagenetTrainable(SupervisedTrainable):
```

Comment:
Maybe `BaseTrainable` and `SupervisedTrainable ` should be on a different file from `SigOptImagenetTrainable` or `SigOptImagenetTrainable` could be renamed to `SigOptSupervisedTrainable `

Reply:
The sipopt trainable uses `SigOptImagenetExperiment` which is quite tied to imagenet experiments (in terms of parameters it assumes we're using). I figured both of these could be made more general, but I've left it for later for now. After Lucas and I do further refactoring, we'll know more which is the best route.